From 21806673ee0daf25f92313d49c8c72edd0a2310b Mon Sep 17 00:00:00 2001
From: maniacx <macs18max@gmail.com>
Date: Sat, 3 Sep 2016 07:36:15 +0800
Subject: [PATCH 2/2] arm: Allow disabling PIE for dynamically linked
 executables

PIE can and does break some binaries that were built without it, resulting in relocation-related issues (addressing memory wrongly assumed to be at a fixed position).

Use this as a last resort measure for compatibility with pre-JB binary blobs, and only if confirmed to fix a broken executable.
If you're not sure, do NOT enable this on a device.

Change-Id: Iddab25f4273077e5a0b02e0be44cdcaeaf6659d2
---
 core/definitions.mk | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/core/definitions.mk b/core/definitions.mk
index 2db4e1b..c31f98b 100644
--- a/core/definitions.mk
+++ b/core/definitions.mk
@@ -92,6 +92,12 @@ $(HOST_2ND_ARCH_VAR_PREFIX)HOST_DEPENDENCIES_ON_SHARED_LIBRARIES :=
 # They are escaped and quoted so can be passed safely to a bash command.
 ANDROID_RESOURCE_GENERATED_CLASSES := 'R.class' 'R$$*.class' 'Manifest.class' 'Manifest$$*.class'
 
+ifeq ($(TARGET_DISABLE_ARM_PIE),true)
+   PIE_EXECUTABLE_TRANSFORM :=
+else
+   PIE_EXECUTABLE_TRANSFORM := -pie
+endif
+
 ###########################################################
 ## Debugging; prints a variable list to stdout
 ###########################################################
@@ -1548,7 +1554,7 @@ endef
 ###########################################################
 
 define transform-o-to-executable-inner
-$(hide) $(PRIVATE_CXX) -pie \
+$(hide) $(PRIVATE_CXX) $(PIE_EXECUTABLE_TRANSFORM) \
 	-nostdlib -Bdynamic \
 	-Wl,-dynamic-linker,$($(PRIVATE_2ND_ARCH_VAR_PREFIX)TARGET_LINKER) \
 	-Wl,--gc-sections \
-- 
1.9.1

